### PR TITLE
migration(democracy): unreserve deposits and clear locks

### PR DESCRIPTION
**Pull request by liamaharon**
Partial https://github.com/paritytech/polkadot-sdk/issues/485

## Overview

Adds a migration that unreserves all deposits and clears locks held by the `democracy` pallet.

I've verified this works using `try-runtime-cli` and the latest Polkadot state (testing with Kusama state blocked by my last Q below).

## Details

`democracy` contains a `DepositOf` storage item containing all the amounts reserved by this pallet, and a `Voters` storage item containing all the accounts that have locks held in the context of this pallet.

`pre_upgrade` collects pre-migration data useful for validating the migration was successful, and also checks the integrity of deposited balances.

`on_runtime_upgrade` unreserves and unlocks the balances.

`post_upgrade` checks that no more locks exist for the pallet, and reserved balances were reduced by the expected amount.

### `try-runtime-cli` `on-runtime-upgrade` logs

```
2023-06-13 16:54:30.693  INFO main runtime::democracy::migrations::unlock_and_unreserve_all_funds: Total accounts: 5364
2023-06-13 16:54:30.693  INFO main runtime::democracy::migrations::unlock_and_unreserve_all_funds: Total stake to unlock: 1179551847104532862
2023-06-13 16:54:30.693  INFO main runtime::democracy::migrations::unlock_and_unreserve_all_funds: Total deposit to unreserve: 25000000000000
2023-06-13 16:54:30.693  INFO main runtime::democracy::migrations::unlock_and_unreserve_all_funds: Bugged deposits: 1/21
```

## Questions for reviewers / todos:

- [x] ~This currently assumes that the `democracy` pallet is part of the runtime. We need to run it for Kusama where it has already been removed from the runtime. What's the cleanest way to go about this?~
- [x] ~How do I calculate the weight of the `remove_lock` and `unreserve` calls?~
- [x] ~`remove_lock` triggers some `ERROR main runtime::system: Logic error: Unexpected underflow in reducing consumer` error logs. I have verified that a non-zero lock always exists before `remove_lock` is called, even when it results in this log. Are they safe to ignore?~

### CODE REVIEW

**Comment by liamaharon** [comment]
I just found a way to handle the Balances types cleaner. Switching this to a draft while I improve that, also going to add some unit tests.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2907933

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2914730

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2914944

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2914950

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2920736

File: frame/democracy/src/migrations/unlock_and_unreserve_all_funds.rs [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+			.count();
+
+		let total_deposits_to_unreserve =
+			account_deposits.clone().into_values().sum::<BalanceOf<T>>();
+		let total_stake_to_unlock = account_locks.clone().into_values().sum::<BalanceOf<T>>();
+
+		log::info!("Total accounts: {:?}", all_accounts.len());
+		log::info!("Total stake to unlock: {:?}", total_stake_to_unlock);
+		log::info!("Total deposit to unreserve: {:?}", total_deposits_to_unreserve);
+		log::info!("Bugged deposits: {}/{}", bugged_deposits, account_deposits.len());
```
Comment (ggwpez) [inline frame/democracy/src/migrations/unlock_and_unreserve_all_funds.rs:155]:
```suggestion
		log::warn!("Bugged deposits: {}/{}", bugged_deposits, account_deposits.len());
```
Maybe also add a  `target` so we know where it came from?

File: frame/democracy/src/migrations/unlock_and_unreserve_all_funds.rs [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+	/// 3. Unlocks the staked funds for each account.
+	fn on_runtime_upgrade() -> frame_support::weights::Weight {
+		// Get staked and deposited balances as reported by this pallet.
+		let (account_deposits, account_stakes, initial_reads) =
+			Self::get_account_deposits_and_locks();
+
+		// Deposited funds need to be unreserved.
+		for (account, unreserve_amount) in account_deposits.iter() {
+			if unreserve_amount.is_zero() {
+				continue
```
Comment (ggwpez) [inline frame/democracy/src/migrations/unlock_and_unreserve_all_funds.rs:174]: Is this expected to be zero?  
If not, also print something please.

File: frame/democracy/src/migrations/unlock_and_unreserve_all_funds.rs [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+
+		// Staked funds need to be unlocked.
+		for account in account_stakes.keys() {
+			T::Currency::remove_lock(DEMOCRACY_ID, account);
+		}
+
+		T::DbWeight::get().reads_writes(
+			account_stakes.len().saturating_add(account_deposits.len()) as u64,
+			account_stakes.len().saturating_add(account_deposits.len()) as u64,
+		) + initial_reads
```
Comment (ggwpez) [inline frame/democracy/src/migrations/unlock_and_unreserve_all_funds.rs:187]:
```suggestion
		).saturating_add(initial_reads)
```

File: frame/democracy/src/migrations/unlock_and_unreserve_all_funds.rs [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+				actual_reserved_before.saturating_sub(expected_amount_deducted);
+			assert!(
+				actual_reserved_after == expected_reserved_after,
+				"Reserved balance for {:?} is incorrect. actual before: {:?}, actual after, {:?}, expected deducted: {:?}",
+				account,
+				actual_reserved_before,
+				actual_reserved_after,
+				expected_amount_deducted,
+			);
+		}
```
Comment (ggwpez) [inline frame/democracy/src/migrations/unlock_and_unreserve_all_funds.rs:223]: Can you also check that the locked amount is zero?

**Review by ggwpez — COMMENTED** [review]
Not super familiar with the pallet but LGTM.  
Going to approve after checking on DotSama myself.

File: frame/democracy/src/migrations/unlock_and_unreserve_all_funds.rs [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+		frame_support::weights::Weight,
+	) {
+		let mut deposit_of_len = 0;
+
+		// Get all deposits (reserved).
+		let mut total_voting_vec_entries: u64 = 0;
+		let account_deposits: BTreeMap<T::AccountId, BalanceOf<T>> = crate::DepositOf::<T>::iter()
+			.flat_map(|(_prop_index, (accounts, balance))| {
+				// Count the number of deposits
+				deposit_of_len.saturating_accrue(1);
```
Comment (kianenigma) [inline frame/democracy/src/migrations/unlock_and_unreserve_all_funds.rs:70]:
```suggestion
				deposit_of_len.saturating_inc();
```

**Review by kianenigma — APPROVED** [review]

**Review by kianenigma — COMMENTED** [review]
Would be good to post an example output of try-runtime prior to final ask for review.

**Review by sam0x17 — APPROVED** [review]

File: frame/democracy/src/migrations/unlock_and_unreserve_all_funds.rs (hunk 4 above)
Comment (liamaharon) [reply frame/democracy/src/migrations/unlock_and_unreserve_all_funds.rs:223]: I originally was, but we can't check the locked amount in `post_upgrade` without adding `pallet-balances` as a regular dependency, after some discussion we decided to remove it and check only check in the tests. See discussion here: https://matrix.to/#/!aJymqQYtCjjqImFLSb:parity.io/$nFYycYWE_wpdiEvt2hGVbmL0euA7rFoF2k1V1pLi8BU?via=parity.io&via=web3.foundation&via=matrix.org

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2980915

**Comment by liamaharon** [comment]
bot merge

**Pull request merged**